In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error
import statsmodels.api as sm
from statsmodels.tools.eval_measures import mse, rmse

import warnings
warnings.filterwarnings('ignore')

In [4]:
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

Önceki derste olduğu gibi, hedef değişkeninizin görünür sıcaklık ve sıcaklık arasındaki fark olduğu bir doğrusal regresyon modeli oluşturun. Açıklayıcı değişkenler olarak nem ve rüzgar hızı kullanın. Şimdi, modelinizi OLS kullanarak tahmin edin. R-kare ve ayarlanmış R-kare değerleri nelerdir? Tatminkar olduklarını düşünüyor musunuz? Açıklayın.

In [6]:
y=weather_df["Temperature (C)"]-weather_df["Apparent Temperature (C)"]

In [7]:
x=weather_df[["Humidity","Wind Speed (km/h)"]]

In [8]:
import statsmodels.api as sm
x = sm.add_constant(x)
results = sm.OLS(y, x).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sat, 17 Apr 2021   Prob (F-statistic):               0.00
Time:                        22:43:25   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R2 çok çok düşük. <0.70 altındaysa kötü bir model. ki bu 0.28

Daha sonra, yukarıdaki modele nem ve rüzgar hızı etkileşimini dahil edin ve OLS'yi kullanarak modeli tahmin edin. Şimdi, bu modelin R-kare değeri nedir? Bu model bir öncekine göre gelişti mi?

In [12]:
weather_df["nem-ruzhız"]=weather_df["Humidity"]*weather_df["Wind Speed (km/h)"]

In [13]:
x=weather_df[["Humidity","Wind Speed (km/h)","nem-ruzhız"]]

In [14]:
import statsmodels.api as sm
x = sm.add_constant(x)
results = sm.OLS(y, x).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sat, 17 Apr 2021   Prob (F-statistic):               0.00
Time:                        22:46:15   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity             -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)    -0.0905      0.002    -36.797      0.000      -0.095      -0.086
nem-ruzhız            0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

0.341 evet bir önceki modele göre gelişti. r2 arttı.

İlk modele ek açıklayıcı değişken olarak görünürlük ekleyin ve tahmin edin. R-kare arttı mı? Ayarlanmış R-kare değeri ne oldu? Tabloda ortaya çıkan farklılıkları, ayarlanmış R-kare içindeki iyileşme açısından etkileşim terimi ve görünürlük ile karşılaştırın. Hangisi daha kullanışlı?

In [15]:
x=weather_df[["Humidity","Wind Speed (km/h)","nem-ruzhız","Visibility (km)"]]

In [16]:
import statsmodels.api as sm
x = sm.add_constant(x)
results = sm.OLS(y, x).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.364
Model:                            OLS   Adj. R-squared:                  0.363
Method:                 Least Squares   F-statistic:                 1.377e+04
Date:                Sat, 17 Apr 2021   Prob (F-statistic):               0.00
Time:                        22:48:35   Log-Likelihood:            -1.6504e+05
No. Observations:               96453   AIC:                         3.301e+05
Df Residuals:                   96448   BIC:                         3.301e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 1.1006      0.039     28.459      0.000       1.025       1.176
Humidity             -0.8909      0.044    -20.263      0.000      -0.977      -0.805
Wind Speed (km/h)    -0.1033      0.002    -42.579      0.000      -0.108      -0.099
nem-ruzhız            0.3164      0.003     95.355      0.000       0.310       0.323
Visibility (km)      -0.0646      0.001    -58.051      0.000      -0.067      -0.062
==============================================================================
Omnibus:                     5328.364   Durbin-Watson:                   0.285
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            11525.074
Skew:                           0.373   Prob(JB):                         0.00
Kurtosis:                       4.520   Cond. No.                         246.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R^2 ler arttı.

AIC ve BIC puanlarına göre yukarıdaki üç modelden en iyisini seçin. Mentor ile gerekçenizi tartışarak seçiminizi doğrulayın.

AIC ve BIC en düşük olan ve r'2 en yüksek olan 3cü modeli seçerdim.

Ev fiyatları verseti ile de lineer regresyonun varsayımlarına bakalım.

In [17]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")

In [18]:
df=house_prices_df[["SalePrice","BldgType","YearBuilt","FullBath","BedroomAbvGr","OverallQual","GarageCars","Street","GrLivArea"]]

In [19]:
X1 = df.drop(['SalePrice'], axis=1)
y1 = np.log(df.SalePrice)
X1=pd.get_dummies(X1)

In [20]:
X1 = sm.add_constant(X1)
result = sm.OLS(y1, X1).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.815
Model:                            OLS   Adj. R-squared:                  0.813
Method:                 Least Squares   F-statistic:                     578.6
Date:                Sat, 17 Apr 2021   Prob (F-statistic):               0.00
Time:                        22:50:57   Log-Likelihood:                 499.07
No. Observations:                1460   AIC:                            -974.1
Df Residuals:                    1448   BIC:                            -910.7
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const               2.7140      0.246     11.051      0.000       2.232       3.196
YearBuilt           0.0031      0.000     14.218      0.000       0.003       0.004
FullBath           -0.0033      0.012     -0.272      0.786      -0.027       0.021
BedroomAbvGr       -0.0192      0.007     -2.621      0.009      -0.034      -0.005
OverallQual         0.1084      0.005     20.214      0.000       0.098       0.119
GarageCars          0.0852      0.008     10.244      0.000       0.069       0.102
GrLivArea           0.0003   1.47e-05     18.929      0.000       0.000       0.000
BldgType_1Fam       0.6352      0.050     12.730      0.000       0.537       0.733
BldgType_2fmCon     0.6199      0.051     12.084      0.000       0.519       0.721
BldgType_Duplex     0.5199      0.054      9.540      0.000       0.413       0.627
BldgType_Twnhs      0.4067      0.057      7.150      0.000       0.295       0.518
BldgType_TwnhsE     0.5324      0.054      9.825      0.000       0.426       0.639
Street_Grvl         1.3094      0.131     10.018      0.000       1.053       1.566
Street_Pave         1.4045      0.125     11.235      0.000       1.159       1.650
==============================================================================
Omnibus:                      637.371   Durbin-Watson:                   2.003
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            10340.008
Skew:                          -1.603   Prob(JB):                         0.00
Kurtosis:                      15.637   Cond. No.                     8.50e+19
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.27e-30. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

R^2 0.81 güzel bir sonuç. Prob (F-statistic):	0.00 p değeri 0. AIC BIC de düşük. Tatmin edici bir sonuç.

In [21]:
X1 = df.drop(['SalePrice'], axis=1)
X1 = X1.drop(['FullBath'], axis=1)
y1 = np.log(df.SalePrice)
X1=pd.get_dummies(X1)

In [22]:
X1 = sm.add_constant(X1)
result = sm.OLS(y1, X1).fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.815
Model:                            OLS   Adj. R-squared:                  0.813
Method:                 Least Squares   F-statistic:                     636.8
Date:                Sat, 17 Apr 2021   Prob (F-statistic):               0.00
Time:                        22:53:10   Log-Likelihood:                 499.03
No. Observations:                1460   AIC:                            -976.1
Df Residuals:                    1449   BIC:                            -917.9
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const               2.7355      0.232     11.775      0.000       2.280       3.191
YearBuilt           0.0031      0.000     14.936      0.000       0.003       0.003
BedroomAbvGr       -0.0196      0.007     -2.718      0.007      -0.034      -0.005
OverallQual         0.1083      0.005     20.269      0.000       0.098       0.119
GarageCars          0.0851      0.008     10.246      0.000       0.069       0.101
GrLivArea           0.0003   1.37e-05     20.207      0.000       0.000       0.000
BldgType_1Fam       0.6400      0.047     13.733      0.000       0.549       0.731
BldgType_2fmCon     0.6239      0.049     12.702      0.000       0.528       0.720
BldgType_Duplex     0.5239      0.052      9.989      0.000       0.421       0.627
BldgType_Twnhs      0.4110      0.055      7.530      0.000       0.304       0.518
BldgType_TwnhsE     0.5368      0.052     10.383      0.000       0.435       0.638
Street_Grvl         1.3206      0.124     10.642      0.000       1.077       1.564
Street_Pave         1.4150      0.119     11.899      0.000       1.182       1.648
==============================================================================
Omnibus:                      634.330   Durbin-Watson:                   2.003
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            10173.112
Skew:                          -1.596   Prob(JB):                         0.00
Kurtosis:                      15.531   Cond. No.                     6.10e+19
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 2.47e-30. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

Fullbath çıkardım ve rkare değişmedi. Ancak AIC BIC değerleri düştü. Bu ikinci modeli kullanabilirim.